server/src/test/java/keywhiz/service/daos/SecretDAOTest.java

```diff
@@ -526,6 +526,148 @@ public void getSecretByIdOneThrowsExceptionIfCurrentVersionIsInvalid() {
+        cryptographer.computeHmac("blah".getBytes(UTF_8), "hmackey"), "creator",
+        ImmutableMap.of(), 0, "", null, null);
+
+    Optional<SecretSeriesAndContent> undeleteSecret3 = secretDAO.getSecretByName("toBeDeletedAndUndeleted");
+    assertThat(undeleteSecret3).isPresent();
+
+    List<SecretSeries> oldDeletedSecrets =
+        secretDAO.getSecretsWithDeletedName("toBeDeletedAndUndeleted");
+    assertThat(oldDeletedSecrets.size()).isEqualTo(2);
+
+    //Because secret Ids are generated randomly, we need to record both secret ids to find the
+    //secret id with two secrets contents associated with it
+    long secretId = oldDeletedSecrets.get(0).id();
+    long otherSecretId = oldDeletedSecrets.get(1).id();
+    Optional<ImmutableList<SecretSeriesAndContent>> secretContents =
+        secretDAO.getDeletedSecretVersionsBySecretId(secretId, 0, 50);
+
+    //If secretId only has 1 secrets content associated, pick the other secret id and retrieve its secrets contents
+    if(secretContents.get().size() == 1) {
+      secretId = otherSecretId;
+      secretContents = secretDAO.getDeletedSecretVersionsBySecretId(secretId, 0, 50);
+    }
+    assertThat(secretContents.isPresent());
+    assertThat(secretContents.get().size()).isEqualTo(2);
+
+    //Arbitrarily pick the 0th secrets content associated with secretId to use in undeleting
+    long contentId = secretContents.get().get(0).content().id();
+
+    assertThat(secretDAO.getSecretByName("toBeDeletedAndUndeleted")).isPresent();
+    secretDAO.updateSecretsCurrent(secretId, contentId);
+    //Because there is already an undeleted secret with the name "toBeDeletedAndUndeleted", we must
+    //rename the deleted secret with a different name
+    secretDAO.renameSecretById(secretId, "toBeDeletedAndUndeleted-2");
+    Optional<SecretSeriesAndContent> undeletedSecretAndContent =
+        secretDAO.getSecretByName("toBeDeletedAndUndeleted-2");
+    assertThat(undeletedSecretAndContent).isPresent();
+    assertThat(undeletedSecretAndContent.get().content().id()).isEqualTo(contentId);
+  }
+
+  //---------------------------------------------------------------------------------------
+  // renameSecret
+  //---------------------------------------------------------------------------------------
+  @Test public void renameSecretById() {
+    long secretId =
+        secretDAO.createSecret("toBeRenamed_renameSecretByIdName", "encryptedShhh",
+        cryptographer.computeHmac("encryptedShhh".getBytes(UTF_8), "hmackey"), "creator",
+        ImmutableMap.of(), 0, "", null, null);
+
+    Optional<SecretSeriesAndContent> secret = secretDAO.getSecretById(secretId);
+    assertThat(secret.get().series().name()).isEqualTo("toBeRenamed_renameSecretByIdName");
+
+    secretDAO.renameSecretById(secretId, "newName");
+
+    secret = secretDAO.getSecretById(secretId);
+    assertThat(secret).isPresent();
+    assertThat(secret.get().series().name()).isEqualTo("newName");
+  }
+
```

Comment:
Let's also double-check that renaming fails if a secret with the same name already exists.